File: src/main/java/com/HelloWay/HelloWay/entities/Product.java
package com.HelloWay.HelloWay.entities;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import lombok.*;

import javax.persistence.*;
import javax.validation.constraints.NotBlank;
import java.util.List;

@Data
@Getter
@Setter
@ToString
@AllArgsConstructor
@Entity
@Table(	name = "products")
@NoArgsConstructor
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "idProduct")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long idProduct;

    @NotBlank
    @Column(length = 50)
    private String productTitle;


    @Column(length = 20)
    private Float price;

    @NotBlank
    @Column(length = 100)
    private String description;

    private Boolean available = true ;

    @OneToMany
    List<BasketProduct> basketProducts;

    @ManyToOne
    @JoinColumn(name = "id_categorie")
    private Categorie categorie;

    @OneToMany(mappedBy="product")
    private List<Image> images;

    public Product(Long idProduct, String productTitle, Float price, String description, Boolean available, Categorie categorie) {
        this.idProduct = idProduct;
        this.productTitle = productTitle;
        this.price = price;
        this.description = description;
        this.available = available;
        this.categorie = categorie;
    }

    public Long getIdProduct() {
        return idProduct;
    }

    public void setIdProduct(Long idProduct) {
        this.idProduct = idProduct;
    }

    public String getProductTitle() {
        return productTitle;
    }

    public void setProductTitle(String productTitle) {
        this.productTitle = productTitle;
    }

    public Float getPrice() {
        return price;
    }

    public void setPrice(Float price) {
        this.price = price;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Boolean getAvailable() {
        return available;
    }

    public void setAvailable(Boolean available) {
        this.available = available;
    }

    public List<BasketProduct> getBasketProducts() {
        return basketProducts;
    }

    public void setBasketProducts(List<BasketProduct> basketProducts) {
        this.basketProducts = basketProducts;
    }

    public Categorie getCategorie() {
        return categorie;
    }

    public void setCategorie(Categorie categorie) {
        this.categorie = categorie;
    }

    public List<Image> getImages() {
        return images;
    }

    public void setImages(List<Image> images) {
        this.images = images;
    }
}

File: fontend/hello_way/lib/models/product.dart


import 'package:hello_way/models/image.dart';

class Product
{
  int? idProduct;
  String productTitle;
  double price;
  String description;
  bool available;
  dynamic categorie;
  List<ImageModel>? images;
  bool? hasActivePromotion;
  int? percentage;
  int? promotionId;
  Product({
    this.idProduct,
    required this.productTitle,
    required this.price,
    required this.description,
    required this.available,
    this.categorie,
    this.images,
    this.hasActivePromotion,
    this.percentage,
    this.promotionId
  });


  factory Product.fromJson(Map<String, dynamic> json) {
    final List<dynamic> jsonImages = json['images'];
    final images = jsonImages.map((image) => ImageModel.fromJson(image)).toList();

    return Product(
        idProduct: json['idProduct'],
        productTitle: json['productTitle'],
        price: json['price'],
        description: json['description'],
        categorie: json['categorie'],
        available: json['available'],
        hasActivePromotion: json['hasActivePromotion'],
        percentage: json['percentage'],
        promotionId: json['promotionId'],
        images: images
    );
  }


  Map<String, dynamic> toJson() => {
    'idProduct': idProduct,
    'productTitle': productTitle,
    'price': price,
    'description': description,
    'categorie': categorie,
    'available':available,
    'hasActivePromotion':hasActivePromotion,
    'percentage':percentage,
    'images': images?.map((image) => image.toJson()).toList(),
  };
}


